x-pack/legacy/plugins/lens/public/indexpattern_plugin/indexpattern_suggestions.tsx

```diff
@@ -0,0 +1,322 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { DatasourceSuggestion } from '../types';
+import { buildColumnForOperationType, getOperationTypesForField } from './operations';
+import { generateId } from '../id_generator';
+import { hasField } from './utils';
+import {
+  IndexPatternPrivateState,
+  IndexPatternField,
+  IndexPatternLayer,
+  columnToOperation,
+} from './indexpattern';
+
+function buildSuggestion(
+  state: IndexPatternPrivateState,
+  layerId: string,
+  isMultiRow: boolean = true,
+  datasourceSuggestionId: number = 0
+) {
+  const columnOrder = state.layers[layerId].columnOrder;
+  const columns = state.layers[layerId].columns;
+  return {
+    state,
+
+    table: {
+      columns: columnOrder.map(columnId => ({
+        columnId,
+        operation: columnToOperation(columns[columnId]),
+      })),
+      isMultiRow,
+      datasourceSuggestionId,
+      layerId,
+    },
+  };
+}
+
+function getIndexPatternIdFromField(
+  state: IndexPatternPrivateState,
+  field: IndexPatternField
+): string | null {
+  if (state.currentIndexPatternId && state.indexPatterns[state.currentIndexPatternId]) {
+    const isMatchingActive = state.indexPatterns[state.currentIndexPatternId].fields.findIndex(f =>
+      _.isEqual(f, field)
+    );
+
+    if (isMatchingActive !== -1) {
+      return state.currentIndexPatternId;
+    }
+  }
+
+  const matchingIndexPattern = Object.values(state.indexPatterns).find(indexPattern => {
+    if (indexPattern.id === state.currentIndexPatternId) {
+      return;
+    }
+
+    const hasMatch = indexPattern.fields.findIndex(f => _.isEqual(f, field));
+
+    if (hasMatch !== -1) {
+      return indexPattern.id;
+    }
+  });
+
+  return matchingIndexPattern ? matchingIndexPattern.id : null;
+}
+
+export function getDatasourceSuggestionsForField(
```

Comment:
I think this function should be limited to just determining which heuristics to apply, and then delegating everything else to various helpers. Something like this:

```
function getSuggestionsForField(state, item) {
  1. If no layers:
    a. Generate new layer
    b. Get suggestions for empty layer (this is the code at the bottom and should be in a function)
  
  2. Find best candidate layer:
    a. Rank based on # of columns (fewer is better) and whether they are already using field
  
  3. For best candidate layer
    a. Determine if we should add a bucket or metric
    b. If adding a bucket, call a helper to add a bucket to layer
    c. If adding a metric, call a function to add metric to layer
```